app/main.py

```diff
@@ -1,6 +1,40 @@
 from typing import Callable
+from functools import wraps
 
 
 def cache(func: Callable) -> Callable:
-    # Write your code here
-    pass
+
+    cache_storage = dict()
+
+    def add_to_storage(func_name: str, args: tuple) -> int:
+
+        func_result = func(*args)
+        cache_storage[func_name].add(args)
+        cache_storage[func_name + "_res"] = {}
+        cache_storage[func_name + "_res"][func_result] = args
```

Comment:
There is a logical error here. You are adding `args` to `cache_storage[func_name]` before initializing `cache_storage[func_name + "_res"]`. This will cause a KeyError if `cache_storage[func_name + "_res"]` does not exist. You should initialize `cache_storage[func_name + "_res"]` before adding `args` to `cache_storage[func_name]`.